cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -600,6 +606,71 @@ public void programSpecification(HttpRequest request, HttpResponder responder,
+   * @param triggerProgramType program type of the triggering program in {@link ProgramStatusTrigger}
+   * @param triggerProgramName program name of the triggering program in {@link ProgramStatusTrigger}
+   * @param triggerProgramStatuses comma separated {@link ProgramStatus} in {@link ProgramStatusTrigger}.
+   *                               Schedules with {@link ProgramStatusTrigger} triggered by none of the
+   *                               {@link ProgramStatus} in triggerProgramStatuses will be filtered out.
+   */
+  @GET
+  @Path("schedules/trigger-type/program-status")
+  public void getProgramStatusSchedules(HttpRequest request, HttpResponder responder,
+                                        @QueryParam("format") String format,
+                                        @QueryParam("trigger-type") String triggerType,
+                                        @QueryParam("trigger-namespace-id") String triggerNamespaceId,
+                                        @QueryParam("trigger-app-name") String triggerAppName,
+                                        @QueryParam("trigger-app-version") String triggerAppVersion,
+                                        @QueryParam("trigger-program-type") String triggerProgramType,
+                                        @QueryParam("trigger-program-name") String triggerProgramName,
+                                        @QueryParam("trigger-program-statuses") String triggerProgramStatuses)
+    throws BadRequestException {
+    boolean asScheduleSpec = returnScheduleAsSpec(format);
+    if (triggerNamespaceId == null && triggerAppName == null && triggerAppVersion == null
+      && triggerProgramType == null && triggerProgramName == null) {
+      throw new BadRequestException("Must specify the triggering program to find schedules " +
+                                      "triggered by program statuses");
+    }
+    // Filter by triggering programId and in program status trigger
+    ApplicationId appId = triggerAppVersion == null ? new ApplicationId(triggerNamespaceId, triggerAppName)
+      : new ApplicationId(triggerNamespaceId, triggerAppName, triggerAppVersion);
+    final ProgramId triggerProgramId = appId.program(ProgramType.valueOfCategoryName(triggerProgramType),
+                                                     triggerProgramName);
+    Collection<ProgramScheduleRecord> schedules = programScheduler.findSchedules(triggerProgramId.toString());
+    if (triggerProgramStatuses != null) {
+      final Set<co.cask.cdap.api.ProgramStatus> queryProgramStatuses = new HashSet<>();
+      for (String status : triggerProgramStatuses.split(",")) {
+        queryProgramStatuses.add(co.cask.cdap.api.ProgramStatus.valueOf(status));
+      }
+      Iterator<ProgramScheduleRecord> iterator = schedules.iterator();
+      while (iterator.hasNext()) {
+        // Remove the schedule if none of its triggering program statuses is in the query program statuses set
```

Comment:
this just saves the step of creating a filtered copy and assign it back to `schedules`. not sure which way is cleaner